filestore/filestore.go

```diff
@@ -129,6 +129,15 @@ func (f *Filestore) Has(c *cid.Cid) (bool, error) {
 }
 
 func (f *Filestore) Put(b blocks.Block) error {
+	has, err := f.Has(b.Cid())
+	if err != nil {
+		return err
+	}
+
+	if has {
+		return nil
+	}
+
 	switch b := b.(type) {
```

Comment:
That becomes problematic, What then happens if i add dataset A, which contains some subset C. Then later, i add dataset B that also contains some subset C. Then i change my mind and remove dataset B. By your description, the original dataset references would now be lost, with no indication to the user that this has happened.

I think this scenario is worse UX than the reverse (deleting A causing bad refs in B).  In either case, repairs are going to have to be manual, and some things will get sticky.

cc @jbenet for his thoughts on this scenario.